### ISSUE TITLE
SSL-related crash in some cases

### ISSUE THREAD

**Issue by redsky17** [Nheko-Reborn/nheko#52]
Some thread dumps during the issue:

https://ocean.joedonofry.com/_matrix/media/v1/download/matrix.org/OCbolWTrxhFzyjKBxcMUURWl

https://ocean.joedonofry.com/_matrix/media/v1/download/matrix.org/ylDUzVbqdciaUMdmxwvgSbPa

https://ocean.joedonofry.com/_matrix/media/v1/download/matrix.org/yGDqVLUaBrKFknBCJVXbCosm

Additional context for the issues:
https://matrix.to/#/!evAqkPHZuPHIGeenhi:ocean.joedonofry.com/$1553636226466434ctUaC:matrix.org?via=ocean.joedonofry.com&via=matrix.org&via=ralith.com

https://matrix.to/#/!evAqkPHZuPHIGeenhi:ocean.joedonofry.com/$1554121096843leWZY:matrix.org?via=ocean.joedonofry.com&via=matrix.org&via=ru-matrix.org

**Comment by MonokelPinguin** [comment]
I've also hit a similar issue, since I'm running my nheko build with the address sanitizer turned on. I actually get a double free in the SSL shutdown, which may be related. This probably belongs to mtxclient and seems to happen, when my PC looses connectivity, although I'm not sure, why my desktop lost connectivity intermittently. Backtrace:

```
[2019-05-03 15:10:07.443] [net] [info] connectivity lost
[2019-05-03 15:10:12.641] [net] [info] trying to re-connect
[2019-05-03 15:44:13.382] [net] [info] connectivity lost
[2019-05-03 15:44:30.407] [net] [info] connectivity lost
[2019-05-03 15:44:42.322] [net] [info] connectivity lost
[2019-05-03 15:44:57.322] [net] [info] connectivity lost
[2019-05-03 15:45:12.321] [net] [info] connectivity lost
[2019-05-03 15:45:27.321] [net] [info] connectivity lost
[2019-05-03 15:46:02.336] [net] [info] connectivity lost
shutdown: uninitialized
[2019-05-03 15:46:22.355] [net] [info] connectivity lost
shutdown: uninitialized
[2019-05-03 15:46:37.363] [net] [info] connectivity lost
shutdown: uninitialized
[2019-05-03 15:46:37.365] [net] [info] connectivity lost
[2019-05-03 15:46:42.702] [net] [info] trying to re-connect
shutdown: unspecified system error
[2019-05-03 15:47:57.322] [net] [info] connectivity lost
[2019-05-03 15:48:12.321] [net] [info] connectivity lost
[2019-05-03 15:48:27.321] [net] [info] connectivity lost
[2019-05-03 15:48:43.334] [net] [info] connectivity lost
[2019-05-03 15:48:57.320] [net] [info] connectivity lost
[2019-05-03 15:49:12.321] [net] [info] connectivity lost
[2019-05-03 15:49:27.322] [net] [info] connectivity lost
[2019-05-03 15:49:42.321] [net] [info] connectivity lost
[2019-05-03 15:49:57.321] [net] [info] connectivity lost
[2019-05-03 15:50:12.321] [net] [info] connectivity lost
[2019-05-03 15:50:27.321] [net] [info] connectivity lost
[2019-05-03 15:50:42.325] [net] [info] connectivity lost
[2019-05-03 15:50:57.659] [net] [info] trying to re-connect
=================================================================
==4706==ERROR: AddressSanitizer: attempting double-free on 0x62900485d200 in thread T2:
    #0 0x7ff22f221810 in __interceptor_free /var/tmp/portage/sys-devel/gcc-8.3.0-r1/work/gcc-8.3.0/libsanitizer/asan/asan_malloc_linux.cc:66
    #1 0x7ff22dedce9b  (/usr/lib64/libssl.so.1.1+0x1ee9b)
    #2 0x7ff22dedaa6e  (/usr/lib64/libssl.so.1.1+0x1ca6e)
    #3 0x7ff22dede592  (/usr/lib64/libssl.so.1.1+0x20592)
    #4 0x7ff22dedbef4  (/usr/lib64/libssl.so.1.1+0x1def4)
    #5 0x7ff22dee1e91  (/usr/lib64/libssl.so.1.1+0x23e91)
    #6 0x7ff22deec0ae in SSL_shutdown (/usr/lib64/libssl.so.1.1+0x2e0ae)
    #7 0x7ff22ec90187 in boost::asio::ssl::detail::engine::do_shutdown(void*, unsigned long) /usr/include/boost/asio/ssl/detail/impl/engine.ipp:313
    #8 0x7ff22ec8fafc in boost::asio::ssl::detail::engine::perform(int (boost::asio::ssl::detail::engine::*)(void*, unsigned long), void*, unsigned long, boost::system::error_code&, unsigned long*) /usr/include/boost/asio/ssl/detail/impl/engine.ipp:235
    #9 0x7ff22ec8eddd in boost::asio::ssl::detail::engine::shutdown(boost::system::error_code&) /usr/include/boost/asio/ssl/detail/impl/engine.ipp:141
    #10 0x7ff22ec90600 in boost::asio::ssl::detail::shutdown_op::operator()(boost::asio::ssl::detail::engine&, boost::system::error_code&, unsigned long&) const /usr/include/boost/asio/ssl/detail/shutdown_op.hpp:37
    #11 0x7ff22ecb2eae in boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >::operator()(boost::system::error_code, unsigned long, int) /usr/include/boost/asio/ssl/detail/io.hpp:145
    #12 0x7ff22ecedc6f in boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>::operator()() /usr/include/boost/asio/detail/bind_handler.hpp:65
    #13 0x7ff22ecf2dfb in void boost::asio::asio_handler_invoke<boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code> >(boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>&, ...) /usr/include/boost/asio/handler_invoke_hook.hpp:69
    #14 0x7ff22ecedafb in void boost_asio_handler_invoke_helpers::invoke<boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >(boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>&, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)>&) /usr/include/boost/asio/detail/handler_invoke_helpers.hpp:37
    #15 0x7ff22ece97e2 in void boost::asio::ssl::detail::asio_handler_invoke<boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>, boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >(boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>&, boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >*) /usr/include/boost/asio/ssl/detail/io.hpp:325
    #16 0x7ff22ece49d3 in void boost_asio_handler_invoke_helpers::invoke<boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>, boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> > >(boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>&, boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >&) /usr/include/boost/asio/detail/handler_invoke_helpers.hpp:37
    #17 0x7ff22ece1ffc in void boost::asio::detail::asio_handler_invoke<boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>, boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>(boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>&, boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>*) /usr/include/boost/asio/detail/bind_handler.hpp:106
    #18 0x7ff22ecdeeeb in void boost_asio_handler_invoke_helpers::invoke<boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>, boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code> >(boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>&, boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>&) /usr/include/boost/asio/detail/handler_invoke_helpers.hpp:37
    #19 0x7ff22ecdca3c in void boost::asio::detail::io_object_executor<boost::asio::executor>::dispatch<boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>, std::allocator<void> >(boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>&&, std::allocator<void> const&) const /usr/include/boost/asio/detail/io_object_executor.hpp:117
    #20 0x7ff22ecd8f28 in void boost::asio::detail::handler_work<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::asio::detail::io_object_executor<boost::asio::executor>, boost::asio::detail::io_object_executor<boost::asio::executor> >::complete<boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code> >(boost::asio::detail::binder1<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::system::error_code>&, boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >&) (/usr/lib64/libmatrix_client.so.0.2.0+0xb43f28)
    #21 0x7ff22ecd2d42 in boost::asio::detail::wait_handler<boost::asio::ssl::detail::io_op<boost::asio::basic_stream_socket<boost::asio::ip::tcp, boost::asio::executor>, boost::asio::ssl::detail::shutdown_op, std::_Bind<void (mtx::http::Session::*(std::shared_ptr<mtx::http::Session>, std::_Placeholder<1>))(boost::system::error_code)> >, boost::asio::detail::io_object_executor<boost::asio::executor> >::do_complete(void*, boost::asio::detail::scheduler_operation*, boost::system::error_code const&, unsigned long) /usr/include/boost/asio/detail/wait_handler.hpp:73
    #22 0x7ff22ea43e82 in boost::asio::detail::scheduler_operation::complete(void*, boost::system::error_code const&, unsigned long) (/usr/lib64/libmatrix_client.so.0.2.0+0x8aee82)
    #23 0x7ff22ea4bfdb in boost::asio::detail::scheduler::do_run_one(boost::asio::detail::conditionally_enabled_mutex::scoped_lock&, boost::asio::detail::scheduler_thread_info&, boost::system::error_code const&) /usr/include/boost/asio/detail/impl/scheduler.ipp:447
    #24 0x7ff22ea4b61a in boost::asio::detail::scheduler::run(boost::system::error_code&) /usr/include/boost/asio/detail/impl/scheduler.ipp:200
    #25 0x7ff22ea4c8e7 in boost::asio::io_context::run() (/usr/lib64/libmatrix_client.so.0.2.0+0x8b78e7)
    #26 0x7ff22ea1fc5f in operator() /var/tmp/portage/dev-libs/mtxclient-9999/work/mtxclient-9999/lib/http/client.cpp:22
    #27 0x7ff22ea3f3b5 in run /usr/include/boost/thread/detail/thread.hpp:120
    #28 0x7ff22e17bc79  (/usr/lib64/libboost_thread.so.1.70.0+0xfc79)
    #29 0x7ff22c842407 in start_thread (/lib64/libpthread.so.0+0x8407)
    #30 0x7ff22c76e7fe in __clone (/lib64/libc.so.6+0x1057fe)

0x62900485d200 is located 0 bytes inside of 16712-byte region [0x62900485d200,0x629004861348)
freed by thread T3 here:
    #0 0x7ff22f221810 in __interceptor_free /var/tmp/portage/sys-devel/gcc-8.3.0-r1/work/gcc-8.3.0/libsanitizer/asan/asan_malloc_linux.cc:66
    #1 0x7ff22dedce9b  (/usr/lib64/libssl.so.1.1+0x1ee9b)

previously allocated by thread T3 here:
    #0 0x7ff22f221bd0 in __interceptor_malloc /var/tmp/portage/sys-devel/gcc-8.3.0-r1/work/gcc-8.3.0/libsanitizer/asan/asan_malloc_linux.cc:86
    #1 0x7ff22dedcc5b  (/usr/lib64/libssl.so.1.1+0x1ec5b)

Thread T2 created by T0 here:
    #0 0x7ff22f185a83 in __interceptor_pthread_create /var/tmp/portage/sys-devel/gcc-8.3.0-r1/work/gcc-8.3.0/libsanitizer/asan/asan_interceptors.cc:202
    #1 0x7ff22e17b228 in boost::thread::start_thread_noexcept() (/usr/lib64/libboost_thread.so.1.70.0+0xf228)

Thread T3 created by T0 here:
    #0 0x7ff22f185a83 in __interceptor_pthread_create /var/tmp/portage/sys-devel/gcc-8.3.0-r1/work/gcc-8.3.0/libsanitizer/asan/asan_interceptors.cc:202
    #1 0x7ff22e17b228 in boost::thread::start_thread_noexcept() (/usr/lib64/libboost_thread.so.1.70.0+0xf228)

SUMMARY: AddressSanitizer: double-free /var/tmp/portage/sys-devel/gcc-8.3.0-r1/work/gcc-8.3.0/libsanitizer/asan/asan_malloc_linux.cc:66 in __interceptor_free
==4706==ABORTING
```

**Comment by dvalter** [comment]
Still persists in 0.7.0-dev (commit dc17358) with the same backtrace

**Comment by deepbluev7** [comment]
It shouldn't, if you update mtxclient to the latest 0.3.0-dev version. This isn't currently done in the bundled deps yet, as I was working on a different branch, but I should probably do that soon...

**Comment by deepbluev7** [comment]
So, this should be fixed with the current 0.7.0-dev versions. I will close this for now, since I can't reproduce it with recent versions anymore. I someone expriences that again (with a recent 0.7.0-dev version), please reopen it.
